### PR TITLE
Implement the weaver-gke purge command.

### PR DESCRIPTION
**Pull request by spetrovic77**
This command will do its best to cleanup the GCP project of all resources created by Service Weaver, including clusters, certificate authorities, IP addresses etc.

Other changes:
  * Rewrite the Certificate Authority code to use Go APIs, so that we can manage its deletion and re-creation more easily.
  * Fix a few bugs in the GCP update code.

### CODE REVIEW

File: internal/gke/cloud_patcher.go [hunk 1]
```diff
@@ -298,6 +302,34 @@ func patchRepository(ctx context.Context, config CloudConfig, opts patchOptions,
...
+	}
+	for r := retry.Begin(); r.Continue(ctx); {
+		if op, err := client.GetOperation(ctx, &containerpb.GetOperationRequest{
+			Name: fmt.Sprintf("projects/%s/locations/%s/operations/%s",
+				config.Project, location, op.Name),
+		}); err == nil && isDone(op) {
+			return doneError
+		}
+	}
+	return fmt.Errorf("timeout waiting for operation %s", op.OperationType)
```
Comment (mwhittaker) [inline internal/gke/cloud_patcher.go:330]: `return fmt.Errorf("timeout waiting for operation %s: %w", op.OperationType, ctx.Err())`?

File: internal/gke/purge.go [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+	if wait, err := deleteConfigCluster(ctx, config); err != nil {
+		errs = append(errs, err)
+	} else if wait != nil {
+		waits = append(waits, wait)
+	}
+	appClusters, err := getApplicationClusters(ctx, config)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	for _, cluster := range appClusters {
```
Comment (mwhittaker) [inline internal/gke/purge.go:71]: I think we shouldn't touch `appClusters` if `getApplicationClusters` returns a non-nil error?

File: internal/gke/purge.go [hunk 3]
```diff
@@ -0,0 +1,791 @@
...
+		if wait, err := deleteCluster(ctx, config, cluster.Name, cluster.Location); err != nil {
+			errs = append(errs, err)
+		} else if wait != nil {
+			waits = append(waits, wait)
+		}
+	}
+
+	// Wait for the clusters to be deleted, as a lot of other resource deletions
+	// cannot proceed while the clusters have a hold on those resources.
+	join()
```
Comment (mwhittaker) [inline internal/gke/purge.go:81]: Should we return after calling `join` if `errs` is non-empty? And maybe before calling `join` too?

File: internal/gke/purge.go [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+		fmt.Fprintf(os.Stderr, "Already deleted\n")
+		return nil
+	}
+	if err != nil {
+		return doErr(err)
+	}
+
+	// Delete the IP address.
+	fmt.Fprintf(os.Stderr, "Deleting... ")
+	_, err = client.Delete(ctx, &computepb.DeleteGlobalAddressRequest{
```
Comment (mwhittaker) [inline internal/gke/purge.go:287]: Do we have to call `Wait` and the returned `Operation`?

File: internal/gke/purge.go [hunk 5]
```diff
@@ -0,0 +1,791 @@
...
+		fmt.Fprintf(os.Stderr, "Already deleted\n")
+		return nil
+	}
+	if err != nil {
+		return doErr(err)
+	}
+
+	// Delete the IP address.
+	fmt.Fprintf(os.Stderr, "Deleting... ")
+	_, err = client.Delete(ctx, &computepb.DeleteAddressRequest{
```
Comment (mwhittaker) [inline internal/gke/purge.go:324]: Same as above, do we need to call `Wait` on the returned `Operation`?

File: internal/gke/purge.go [hunk 6]
```diff
@@ -0,0 +1,791 @@
...
+		return nil
+	}
+
+	// Delete DNS record sets inside the DNS zone, as the zone must be empty
+	// before it can be deleted.
+	records, err := getDNSRecordSets(ctx, config, zone)
+	if err != nil {
+		return nil
+	}
+	if len(records) == 0 { // already deleted
```
Comment (mwhittaker) [inline internal/gke/purge.go:570]: I'm not familiar with DNS zones, but is it possible to have a DNS zone without any records?

File: internal/gke/purge.go [hunk 7]
```diff
@@ -0,0 +1,791 @@
...
+	fmt.Fprintf(os.Stderr, "Fetching custom role %q... ", role)
+	iamService, err := iam.NewService(ctx, config.ClientOptions()...)
+	if err != nil {
+		return doErr(err)
+	}
+	service := iam.NewProjectsService(iamService).Roles
+	getCall := service.Get(path)
+	getCall.Context(ctx)
+	r, err := getCall.Do()
+	if isNotFound(err) || r.Deleted {
```
Comment (mwhittaker) [inline internal/gke/purge.go:350]: Do we need to check that `err` is not-nil before touching `r.Deleted`?

File: internal/gke/purge.go [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+	switch ca.State {
+	case privatecapb.CertificateAuthority_DELETED:
+		// Already deleted (but not permanently): nothing to do but wait 30 days.
+		return nil, nil
+	case privatecapb.CertificateAuthority_ENABLED:
+		// Enabled: must disable before deleting.
+		if err := disableCA(ctx, config, path); err != nil {
+			return nil, err
+		}
+		fallthrough
```
Comment (mwhittaker) [inline internal/gke/purge.go:429]: Woah, I never knew this was a thing.

**Review by mwhittaker — APPROVED** [review]
Super super impressive stuff! I don't know how you learn how to use all the various cloud APIs so quickly. :brain: 

Love the branch name too :laughing:

**Comment by spetrovic77** [comment]
> Super super impressive stuff! I don't know how you learn how to use all the various cloud APIs so quickly. 🧠
> 
> Love the branch name too 😆

Thanks for the thorough review. I went through the cloud patcher code again and made some RMW changes.

**Pull request merged**